### PR TITLE
Feat/forgot password

### PR DESCRIPTION
**Pull request by PVCuong2001**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -21,23 +21,27 @@
     "npm": "8.11.0"
   },
   "dependencies": {
+    "bcrypt": "^5.1.0",
```
Comment (minhhoangho) [inline package.json:24]: Please use fixed version of dependencies to avoid conflict version of relative deps

File: src/auth/router.js [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+const bcrypt = require('bcrypt');
+const jsonwebtoken = require('jsonwebtoken');
+const crypto = require('crypto');
+
+const db = require('../database/connection');
+const sendEmail = require('../utils/emailhelper');
+
+const router = express.Router();
+
+const executeQuery = ({ query, params }) => {
```
Comment (minhhoangho) [inline src/auth/router.js:11]: This function could move to common/base file for reusable purpose

File: src/auth/router.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  return new Promise((resolve, reject) => {
+    db.query(query, params, (err, rows) => {
+      if (err) {
+        reject(err);
+      }
+
+      if (rows.length === 0) {
+        resolve(null);
```
Comment (minhhoangho) [inline src/auth/router.js:19]: Please consider for query the case get list (return empty array) and get one (return null)

File: src/auth/router.js [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+      }
+
+      resolve(rows[0]);
```
Comment (minhhoangho) [inline src/auth/router.js:22]: I think this is quite confused when we want to get list of data, right?

For this function, I think we can put more parameters of query type (getMany, getOne, createOne, createMany, ...) or something like that. So we can reuse this function effectively.

**Review by minhhoangho — COMMENTED** [review]
I left some comments here, PTAL

File: src/auth/router.js [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+      params: [email],
+    });
+
+    if (!user) {
+      return res.status(404).json({
+        message: 'Email not found',
+      });
+    }
+
+    const resetToken = crypto.randomBytes(32).toString('hex');
```
Comment (dangphu2412) [inline src/auth/router.js:97]: This is secret key instead of resetToken

File: src/auth/router.js [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    const passwordResetToken = crypto.createHash('sha256').update(resetToken).digest('hex');
+
+    const passwordResetAt = new Date(Date.now() + 10 * 60 * 1000);
+    await executeQuery({
+      query: 'update user set passwordResetToken = ?, passwordResetAt = ? where email = ?',
+      params: [passwordResetToken, passwordResetAt, email],
+    });
```
Comment (dangphu2412) [inline src/auth/router.js:104]: Pls rebase the new code and resolve conflict

File: src/auth/router.js [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+    sendEmail({
+      emailFrom: 'admin@gmail.com',
+      emailTo: email,
+      emailSubject: 'Reset password',
+      emailText: 'Here is your reset password token: ' + passwordResetToken,
+    });
```
Comment (dangphu2412) [inline src/auth/router.js:111]: move sendEmail utils to services/mail.service.js
where contains mailService like:
```
const mailService = {
   send() {}
};

Object.freeze(mailService)

module.exports = {
   mailService
}
```

**Pull request closed without merging**